File: paos/crates/paos-web/src/tasks.js
/* The kanban board.
 *
 * Dependency-free, and textContent everywhere — NEVER innerHTML. Every string rendered
 * here (titles, bodies, note text, handles) is arbitrary text written by other sessions,
 * so the safe primitive is the only primitive. Same rule as index.html, same reason.
 *
 * Exposes one global, `TasksView`, which index.html calls when its nav lands on 'tasks'.
 */
(function () {
  const COLUMNS = [
    ['proposed', 'Proposed'],
    ['ready', 'Ready'],
    ['in_progress', 'In progress'],
    ['review', 'Review'],
    ['done', 'Done'],
  ];

  const el = (tag, cls, text) => {
    const n = document.createElement(tag);
    if (cls) n.className = cls;
    if (text !== undefined) n.textContent = text;
    return n;
  };

  const state = {
    data: null,
    filters: { repo: '', owner: '', scope: '', orphaned: false, q: '' },
    swimlanes: false,
    // Everything that must survive a poll. A refresh that discarded a half-typed comment
    // or cancelled a drag is how a live board becomes one people close.
    dragging: null,
    openId: null,
    dirty: false,
    timer: null,
  };

  function readHash() {
    const h = new URLSearchParams((location.hash || '').replace(/^#/, ''));
    if (h.get('view') !== 'tasks') return;
    state.filters.repo = h.get('repo') || '';
    state.filters.owner = h.get('owner') || '';
    state.filters.scope = h.get('scope') || '';
    state.filters.orphaned = h.get('orphaned') === '1';
    state.filters.q = h.get('q') || '';
    state.swimlanes = h.get('lanes') === '1';
  }

  /** Filters live in the URL so a filtered board is a link you can send yourself. */
  function writeHash() {
    const p = new URLSearchParams({ view: 'tasks' });
    const f = state.filters;
    if (f.repo) p.set('repo', f.repo);
    if (f.owner) p.set('owner', f.owner);
    if (f.scope) p.set('scope', f.scope);
    if (f.orphaned) p.set('orphaned', '1');
    if (f.q) p.set('q', f.q);
    if (state.swimlanes) p.set('lanes', '1');
    history.replaceState(null, '', '#' + p.toString());
  }

  async function api(path, opts) {
    const r = await fetch(path, opts);
    if (!r.ok) throw new Error((await r.text()) || r.status);
    return r.json();
  }

  const post = (path, data) =>
    api(path, {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: new URLSearchParams(data).toString(),
    });

  function toast(msg) {
    const t = el('div', 'toast', msg);
    document.body.append(t);
    setTimeout(() => t.remove(), 4200);
  }

  function visible(t) {
    const f = state.filters;
    if (f.repo && t.repo !== f.repo) return false;
    if (f.owner && t.claimed_by !== f.owner) return false;
    if (f.scope && t.scope !== f.scope) return false;
    if (f.orphaned && !(t.orphaned && t.unowned)) return false;
    if (f.q) {
      const hay = (t.title + ' ' + (t.body || '') + ' ' + t.id).toLowerCase();
      if (!hay.includes(f.q.toLowerCase())) return false;
    }
    return true;
  }

  // ---- cards ---------------------------------------------------------------

  function card(t) {
    const c = el('div', 'card' + (t.rescue ? ' rescue' : ''));
    c.draggable = true;
    c.dataset.id = t.id;

    c.addEventListener('dragstart', (e) => {
      state.dragging = t.id;
      c.classList.add('dragging');
      e.dataTransfer.setData('text/plain', t.id);
      e.dataTransfer.effectAllowed = 'move';
    });
    c.addEventListener('dragend', () => {
      state.dragging = null;
      c.classList.remove('dragging');
    });
    c.addEventListener('click', (e) => {
      if (e.target.tagName === 'SELECT') return;
      openDetail(t.id);
    });

    c.append(el('div', 't', t.title));

    const meta = el('div', 'meta');
    meta.append(el('span', 'pill p' + t.priority, 'p' + t.priority));
    if (t.rescue) {
      meta.append(el('span', 'pill rescue', '⤺ rescue' + (t.last_owner ? ' · ' + t.last_owner : '')));
    } else if (t.claimed_by) {
      meta.append(el('span', 'pill owner', '@' + t.claimed_by));
    } else if (t.unowned && t.last_owner) {
      meta.append(el('span', 'pill', 'released'));
    }
    if (t.blocked) meta.append(el('span', 'pill blocked', '⛔ blocked'));
    if (t.repo) meta.append(el('span', 'pill', t.repo));
    if (t.notes) meta.append(el('span', 'pill', '💬 ' + t.notes));

    // The accessible path, and the only one that works on a phone — which is where the
    // operator answers from. Drag is the flourish; this is the interface.
    const sel = el('select');
    for (const [v, label] of COLUMNS) {
      const o = el('option', null, label);
      o.value = v;
      if (v === t.state) o.selected = true;
      sel.append(o);
    }
    const dropped = el('option', null, 'Dropped');
    dropped.value = 'dropped';
    if (t.state === 'dropped') dropped.selected = true;
    sel.append(dropped);
    sel.onchange = () => move(t.id, sel.value, t.state);
    meta.append(sel);

    c.append(meta);
    return c;
  }

  /** Optimistic, with a rollback. The SERVER decides whether a move is legal. */
  async function move(id, to, from) {
    if (to === from) return;
    try {
      await post('/api/task/state', { id, to });
      await load();
    } catch (e) {
      toast(String(e.message || e));
      await load(); // snap the card back to whatever the server actually thinks
    }
  }

  // ---- board ---------------------------------------------------------------

  function boardFor(tasks) {
    const board = el('div', 'board');
    for (const [key, label] of COLUMNS) {
      const col = el('div', 'col');
      const mine = tasks.filter((t) => t.state === key);
      const h = el('h3');
      h.append(el('span', null, label), el('span', 'n', String(mine.length)));
      col.append(h);

      const drop = el('div', 'drop');
      drop.addEventListener('dragover', (e) => {
        e.preventDefault();
        drop.classList.add('over');
      });
      drop.addEventListener('dragleave', () => drop.classList.remove('over'));
      drop.addEventListener('drop', (e) => {
        e.preventDefault();
        drop.classList.remove('over');
        const id = e.dataTransfer.getData('text/plain') || state.dragging;
        const t = state.data.tasks.find((x) => x.id === id);
        if (t) move(id, key, t.state);
      });
      for (const t of mine) drop.append(card(t));
      col.append(drop);
      board.append(col);
    }
    return board;
  }

  function render(root) {
    root.replaceChildren();
    if (!state.data) {
      root.append(el('div', 'empty', 'loading…'));
      return;
    }
    root.append(controls());

    const tasks = state.data.tasks.filter(visible);
    if (!tasks.length) {
      root.append(el('div', 'empty', 'no tasks match — clear the filters, or create one above'));
      return;
    }

    if (!state.swimlanes) {
      root.append(boardFor(tasks));
    } else {
      const byId = new Map(state.data.tasks.map((t) => [t.id, t]));
      const epics = new Map();
      const loose = [];
      for (const t of tasks) {
        if (t.parent_id && byId.has(t.parent_id)) {
          if (!epics.has(t.parent_id)) epics.set(t.parent_id, []);
          epics.get(t.parent_id).push(t);
        } else {
          loose.push(t);
        }
      }
      for (const [pid, kids] of epics) {
        root.append(lane(byId.get(pid).title, kids, pid));
      }
      if (loose.length) root.append(lane('Unparented', loose, '_loose'));
    }
    if (state.openId) openDetail(state.openId, true);
  }

  function lane(title, tasks, key) {
    const wrap = el('div', 'lane');
    const open = sessionStorage.getItem('lane:' + key) !== 'closed';
    wrap.dataset.open = String(open);
    const h = el('h2');
    h.append(el('span', 'caret', open ? '▼' : '▶'), el('span', null, title),
             el('span', 'n', String(tasks.length)));
    h.onclick = () => {
      const now = wrap.dataset.open !== 'true';
      wrap.dataset.open = String(now);
      sessionStorage.setItem('lane:' + key, now ? 'open' : 'closed');
      h.firstChild.textContent = now ? '▼' : '▶';
    };
    wrap.append(h, boardFor(tasks));
    return wrap;
  }

  function controls() {
    const bar = el('div', 'board-bar');

    const add = el('input');
    add.placeholder = 'new task — type a title and press Enter';
    add.onkeydown = async (e) => {
      if (e.key !== 'Enter' || !add.value.trim()) return;
      try {
        await post('/api/task/create', {
          title: add.value.trim(),
          scope: state.filters.repo ? 'project' : 'global',
          repo: state.filters.repo || '',
        });
        add.value = '';
        await load();
      } catch (err) {
        toast(String(err.message || err));
      }
    };
    bar.append(add);

    const pick = (label, key, values) => {
      const s = el('select');
      const any = el('option', null, label);
      any.value = '';
      s.append(any);
      for (const v of values) {
        const o = el('option', null, v);
        o.value = v;
        if (state.filters[key] === v) o.selected = true;
        s.append(o);
      }
      s.onchange = () => {
        state.filters[key] = s.value;
        writeHash();
        render(document.getElementById('app'));
      };
      return s;
    };
    bar.append(pick('any repo', 'repo', state.data.repos));
    bar.append(pick('anyone', 'owner', state.data.owners));
    bar.append(pick('any scope', 'scope', ['global', 'org', 'project']));

    const orph = el('button', 'chip', '⤺ unowned only');
    orph.setAttribute('aria-pressed', String(state.filters.orphaned));
    orph.onclick = () => {
      state.filters.orphaned = !state.filters.orphaned;
      writeHash();
      render(document.getElementById('app'));
    };
    bar.append(orph);

    const lanes = el('button', 'chip', 'swimlanes');
    lanes.setAttribute('aria-pressed', String(state.swimlanes));
    lanes.onclick = () => {
      state.swimlanes = !state.swimlanes;
      writeHash();
      render(document.getElementById('app'));
    };
    bar.append(lanes);

    const q = el('input');
    q.placeholder = 'filter…';
    q.value = state.filters.q;
    q.style.maxWidth = '160px';
    q.oninput = () => {
      state.filters.q = q.value;
      writeHash();
      const app = document.getElementById('app');
      const scroll = app.scrollTop;
      render(app);
      app.scrollTop = scroll;
      const again = document.querySelector('.board-bar input:last-of-type');
      if (again) { again.focus(); again.setSelectionRange(q.value.length, q.value.length); }
    };
    bar.append(el('span', 'spacer'), q);
    return bar;
  }

  // ---- detail panel --------------------------------------------------------

  async function openDetail(id, silent) {
    state.openId = id;
    let d;
    try {
      d = await api('/api/task?id=' + encodeURIComponent(id));
    } catch (e) {
      if (!silent) toast(String(e.message || e));
      state.openId = null;
      return;
    }
    document.querySelectorAll('.panel, .backdrop').forEach((n) => n.remove());

    const back = el('div', 'backdrop');
    back.onclick = closeDetail;
    const p = el('div', 'panel');
    const t = d.task;

    const x = el('button', 'act close', 'close');
    x.onclick = closeDetail;
    p.append(x);

    p.append(el('div', 'k', t.id + '  ·  ' + t.state + '  ·  p' + t.priority));
    p.append(el('h3', null, t.title));

    // Why it is unowned goes ABOVE the body: someone picking this up needs to know what
    // they are inheriting before they read what the work is.
    if (t.unowned && t.state !== 'done' && t.state !== 'dropped' && t.last_owner) {
      p.append(el('div', 'why', t.orphaned
        ? '⤺ Unowned — ' + t.last_owner + ' ended while holding this. Open to rescue.'
        : '⤺ Unowned — released by ' + t.last_owner + '.'));
    }

    const meta = [];
    meta.push('scope ' + t.scope);
    if (t.repo) meta.push('repo ' + t.repo);
    if (t.claimed_by) meta.push('held by ' + t.claimed_by);
    meta.push('created by ' + t.created_by + ' (' + t.origin + ')');
    if (t.close_grant) meta.push('sessions may close this');
    p.append(el('div', 'fix', meta.join(' · ')));

    if (t.body) p.append(el('div', 'txt', t.body));

    if (d.deps.length) {
      p.append(el('h2', null, 'depends on'));
      for (const dep of d.deps) {
        const row = el('div', 'fix', dep.id + ' [' + dep.state + '] ' + dep.title);
        p.append(row);
      }
    }

    if (d.notes.length) {
      p.append(el('h2', null, 'log'));
      const log = el('div', 'log');
      for (const n of d.notes) {
        const e = el('div', 'entry ' + (n.kind || 'note'));
        e.append(el('div', 'who', n.ts + '  ' + n.author));
        e.append(el('div', 'txt', n.text));
        log.append(e);
      }
      p.append(log);
    }

    p.append(el('h2', null, 'comment'));
    const box = el('textarea');
    box.placeholder = 'a note on this task…';
    box.oninput = () => { state.dirty = box.value.trim().length > 0; };
    p.append(box);

    const wakeWrap = el('label');
    const wake = el('input');
    wake.type = 'checkbox';
    wake.checked = !!t.claimed_by;
    wake.style.width = 'auto';
    wakeWrap.append(wake, document.createTextNode(
      t.claimed_by ? 'wake ' + t.claimed_by : 'wake the owner (nobody holds this)'));
    p.append(wakeWrap);

    const acts = el('div', 'acts');
    const send = el('button', 'act primary', 'comment');
    send.onclick = async () => {
      if (!box.value.trim()) return;
      try {
        const r = await post('/api/task/note', {
          id: t.id, text: box.value.trim(), ...(wake.checked ? { wake: '1' } : {}),
        });
        box.value = '';
        state.dirty = false;
        toast(r.delivered ? 'commented and woke ' + t.claimed_by
                          : 'commented — ' + (r.why || 'not delivered'));
        openDetail(t.id, true);
      } catch (e) {
        toast(String(e.message || e));
      }
    };
    acts.append(send);

    if (t.origin === 'operator' && !t.close_grant) {
      const g = el('button', 'act', 'let a session close this');
      g.onclick = async () => {
        try {
          await post('/api/task/grant', { id: t.id });
          toast('granted');
          openDetail(t.id, true);
          load();
        } catch (e) { toast(String(e.message || e)); }
      };
      acts.append(g);
    }
    p.append(acts);

    document.body.append(back, p);
    if (!silent) box.focus();
  }

  function closeDetail() {
    state.openId = null;
    state.dirty = false;
    document.querySelectorAll('.panel, .backdrop').forEach((n) => n.remove());
  }

  // ---- lifecycle -----------------------------------------------------------

  async function load() {
    state.data = await api('/api/tasks');
    const badge = document.getElementById('tasks-badge');
    if (badge) {
      badge.textContent = state.data.needs_operator || '';
      badge.style.display = state.data.needs_operator ? '' : 'none';
    }
    render(document.getElementById('app'));
  }

  const TasksView = {
    async show() {
      readHash();
      writeHash();
      await load();
      clearInterval(state.timer);
      // Poll, but never on top of the user. A drag in flight or a half-typed comment is
      // work in progress, and clobbering it costs more than five seconds of staleness.
      state.timer = setInterval(() => {
        if (state.dragging || state.dirty) return;
        load().catch(() => {});
      }, 5000);
    },
    hide() {
      clearInterval(state.timer);
      state.timer = null;
      closeDetail();
    },
    async badge() {
      try {
        const d = await api('/api/tasks');
        return d.needs_operator || 0;
      } catch (e) {
        return 0;
      }
    },
  };

  window.TasksView = TasksView;
  document.addEventListener('keydown', (e) => {
    if (e.key === 'Escape' && state.openId) closeDetail();
  });
})();

File: paos/crates/paos-web/src/tasks.rs
//! The board's server side.
//!
//! Split out of `lib.rs` rather than appended to it: that file was already carrying
//! every other view, and a kanban board's payload plus five write routes is not a
//! footnote. Same crate, so `impl Web` here is the same type — only the file is smaller.
//!
//! The dashboard always acts as the OPERATOR. That is what makes the board able to
//! approve `review` items, and it is why `may_close` is consulted rather than assumed:
//! the policy lives in `paos-tasks` and neither caller reimplements it.

use crate::http::{Request, Response};
use crate::{field, now_iso, Web};
use paos_tasks::model::{Origin, State, Task};
use paos_tasks::{query, store};

const TASKS_CSS: &str = include_str!("tasks.css");
const TASKS_JS: &str = include_str!("tasks.js");

pub fn asset(path: &str) -> Option<Response> {
    match path {
        "/tasks.css" => Some(Response::css(TASKS_CSS)),
        "/tasks.js" => Some(Response::js(TASKS_JS)),
        _ => None,
    }
}

impl Web {
    /// The whole board in one payload.
    ///
    /// Three queries and a join in Rust, rather than a correlated subquery per card: the
    /// board renders a few hundred rows at once, and per-row `EXISTS` on a table this
    /// small is the wrong shape for it.
    pub(crate) fn tasks(&self) -> Response {
        let c = self.lock();
        let tasks = match query::all(&c) {
            Ok(t) => t,
            Err(e) => return Response::text(500, &e),
        };
        let blocked = query::blocked_ids(&c).unwrap_or_default();
        let notes = query::note_counts(&c).unwrap_or_default();
        let needs = query::needs_operator(&c).unwrap_or(0);

        let items: Vec<String> = tasks
            .iter()
            .map(|t| task_json(t, blocked.contains(&t.id), *notes.get(&t.id).unwrap_or(&0)))
            .collect();

        // Repos and scopes the board can filter by, taken from the data rather than a
        // hardcoded list — a filter offering a repo with no tasks is a dead end.
        let mut repos: Vec<&str> = tasks.iter().filter_map(|t| t.repo.as_deref()).collect();
        repos.sort_unstable();
        repos.dedup();
        let mut owners: Vec<&str> = tasks
            .iter()
            .filter_map(|t| t.claimed_by.as_deref())
            .collect();
        owners.sort_unstable();
        owners.dedup();

        Response::json(format!(
            "{{\"tasks\":[{}],\"needs_operator\":{needs},\"repos\":[{}],\"owners\":[{}]}}",
            items.join(","),
            repos.iter().map(|r| jstr(r)).collect::<Vec<_>>().join(","),
            owners.iter().map(|o| jstr(o)).collect::<Vec<_>>().join(","),
        ))
    }

    /// One task, with the full log — what the detail panel opens.
    pub(crate) fn task_one(&self, req: &Request) -> Response {
        let Some(id) = req.query.get("id") else {
            return Response::text(400, "missing id");
        };
        let c = self.lock();
        let t = match store::get(&c, id) {
            Ok(Some(t)) => t,
            Ok(None) => return Response::not_found(),
            Err(e) => return Response::text(500, &e),
        };
        let blocked = query::is_blocked(&c, id).unwrap_or(false);
        let notes: Vec<String> = query::notes(&c, id)
            .unwrap_or_default()
            .iter()
            .map(|n| {
                format!(
                    "{{\"ts\":{},\"author\":{},\"kind\":{},\"text\":{}}}",
                    jstr(&n.ts),
                    jstr(&n.author),
                    jstr(&n.kind),
                    jstr(&n.text)
                )
            })
            .collect();
        let deps: Vec<String> = query::deps(&c, id)
            .unwrap_or_default()
            .iter()
            .map(|(did, dstate, dtitle)| {
                format!(
                    "{{\"id\":{},\"state\":{},\"title\":{}}}",
                    jstr(did),
                    jstr(dstate.as_str()),
                    jstr(dtitle)
                )
            })
            .collect();
        Response::json(format!(
            "{{\"task\":{},\"notes\":[{}],\"deps\":[{}]}}",
            task_json(&t, blocked, notes.len() as i64),
            notes.join(","),
            deps.join(",")
        ))
    }

    pub(crate) fn task_create(&self, req: &Request) -> Response {
        let Some(title) = field(&req.body, "title").filter(|t| !t.trim().is_empty()) else {
            return Response::text(400, "a task needs a title");
        };
        let scope = field(&req.body, "scope").unwrap_or_else(|| "global".into());
        let opt = |k: &str| field(&req.body, k).filter(|v| !v.trim().is_empty());
        let n = paos_tasks::model::NewTask {
            title: title.trim().to_string(),
            body: opt("body"),
            scope,
            org: opt("org"),
            repo: opt("repo"),
            parent_id: opt("parent_id"),
            priority: opt("priority").and_then(|p| p.parse().ok()).unwrap_or(2),
            // Created from the dashboard means created by the operator. That is what
            // decides who may close it, so it is not a cosmetic label.
            origin: Origin::Operator,
            created_by: "operator".into(),
            room: opt("room"),
            start_ready: opt("start_ready").is_some(),
        };
        let c = self.lock();
        match store::create(&c, &n, &now_iso()) {
            Ok(id) => Response::json(format!("{{\"ok\":true,\"id\":{}}}", jstr(&id))),
            Err(e) => Response::text(400, &e),
        }
    }

    /// Move a card. The server is the authority on whether the move is legal — the board
    /// animates optimistically and rolls back on a non-200.
    pub(crate) fn task_state(&self, req: &Request) -> Response {
        let Some(id) = field(&req.body, "id") else {
            return Response::text(400, "missing id");
        };
        let Some(to) = field(&req.body, "to").and_then(|s| State::parse(&s)) else {
            return Response::text(400, "unknown state");
        };
        let c = self.lock();
        match store::set_state(&c, &id, to, &store::Actor::Operator, &now_iso()) {
            Ok(()) => Response::json("{\"ok\":true}".into()),
            // 409, not 500: a refused transition is an answer, and the board needs to
            // tell them apart to know whether to show the reason or a generic failure.
            Err(e) => Response::text(409, &e),
        }
    }

    /// Comment on a task, optionally waking whoever is working it.
    ///
    /// The note is written FIRST and unconditionally. A comment must never be lost
    /// because the wake failed — a board you can only watch is a status page, and the
    /// point of this route is that it is not one.
    pub(crate) fn task_note(&self, req: &Request) -> Response {
        let Some(id) = field(&req.body, "id") else {
            return Response::text(400, "missing id");
        };
        let Some(text) = field(&req.body, "text").filter(|t| !t.trim().is_empty()) else {
            return Response::text(400, "empty comment");
        };
        let (owner, room) = {
            let c = self.lock();
            match store::get(&c, &id) {
                Ok(Some(t)) => {
                    if let Err(e) =
                        store::note(&c, &id, "operator", "comment", text.trim(), &now_iso())
                    {
                        return Response::text(500, &e);
                    }
                    (t.claimed_by.clone(), t.room.clone())
                }
                Ok(None) => return Response::not_found(),
                Err(e) => return Response::text(500, &e),
            }
        };
        if field(&req.body, "wake").is_none() {
            return Response::json("{\"ok\":true,\"delivered\":false}".into());
        }
        let Some(owner) = owner else {
            return Response::json(
                "{\"ok\":true,\"delivered\":false,\"why\":\"nobody is holding this task\"}".into(),
            );
        };
        // The note lives on the task, so the wake only has to be a pointer to it.
        let head: String = text.trim().chars().take(80).collect();
        let body = format!("task {id}: {head}");
        let room = room.unwrap_or_else(|| "lobby".into());
        match crate::run_skill(&["bus", "wake", &format!("@{owner}"), &body, "--room", &room]) {
            Ok(_) => Response::json("{\"ok\":true,\"delivered\":true}".into()),
            Err(e) => Response::json(format!(
                "{{\"ok\":true,\"delivered\":false,\"why\":{}}}",
                jstr(&e)
            )),
        }
    }

    pub(crate) fn task_grant(&self, req: &Request) -> Response {
        let Some(id) = field(&req.body, "id") else {
            return Response::text(400, "missing id");
        };
        let c = self.lock();
        match store::grant_close(&c, &id, &now_iso()) {
            Ok(()) => Response::json("{\"ok\":true}".into()),
            Err(e) => Response::text(400, &e),
        }
    }

    pub(crate) fn task_dep(&self, req: &Request) -> Response {
        let (Some(id), Some(dep)) = (field(&req.body, "id"), field(&req.body, "depends_on")) else {
            return Response::text(400, "missing id or depends_on");
        };
        let remove = field(&req.body, "remove").is_some();
        let c = self.lock();
        let r = if remove {
            store::dep_rm(&c, &id, &dep)
        } else {
            store::dep_add(&c, &id, &dep, &now_iso())
        };
        match r {
            Ok(()) => Response::json("{\"ok\":true}".into()),
            Err(e) => Response::text(409, &e),
        }
    }
}

fn task_json(t: &Task, blocked: bool, notes: i64) -> String {
    format!(
        "{{\"id\":{},\"title\":{},\"body\":{},\"state\":{},\"priority\":{},\"scope\":{},\
         \"org\":{},\"repo\":{},\"parent_id\":{},\"origin\":{},\"created_by\":{},\
         \"claimed_by\":{},\"last_owner\":{},\"orphaned\":{},\"close_grant\":{},\"room\":{},\
         \"blocked\":{},\"notes\":{},\"unowned\":{},\"rescue\":{},\"created_ts\":{},\
         \"updated_ts\":{}}}",
        jstr(&t.id),
        jstr(&t.title),
        ostr(t.body.as_deref()),
        jstr(t.state.as_str()),
        t.priority,
        jstr(&t.scope),
        ostr(t.org.as_deref()),
        ostr(t.repo.as_deref()),
        ostr(t.parent_id.as_deref()),
        jstr(t.origin.as_str()),
        jstr(&t.created_by),
        ostr(t.claimed_by.as_deref()),
        ostr(t.last_owner.as_deref()),
        t.orphaned,
        t.close_grant,
        ostr(t.room.as_deref()),
        blocked,
        notes,
        t.is_unowned(),
        t.is_rescue(),
        jstr(&t.created_ts),
        jstr(&t.updated_ts),
    )
}

/// A quoted, escaped JSON string.
///
/// Uses the crate's own `esc` rather than serde_json, which paos-web deliberately does
/// not depend on. Task titles and note text are arbitrary strings written by other
/// sessions, so a quote or a newline in one must not be able to break the payload —
/// which is exactly what `esc` exists and is tested for.
fn jstr(s: &str) -> String {
    format!("\"{}\"", crate::http::esc(s))
}

fn ostr(s: Option<&str>) -> String {
    s.map(jstr).unwrap_or_else(|| "null".into())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::{Arc, Mutex};

    fn test_web() -> Web {
        let c = paos_store::open_in_memory().unwrap();
        paos_memory::ensure_schema(&c).unwrap();
        Web {
            conn: Arc::new(Mutex::new(c)),
            embedder: Arc::new(paos_memory::HashEmbedder::new(64)),
        }
    }

    fn get(path: &str) -> Request {
        let (p, q) = match path.split_once('?') {
            Some((p, q)) => (p, q),
            None => (path, ""),
        };
        let mut query = std::collections::HashMap::new();
        for pair in q.split('&').filter(|s| !s.is_empty()) {
            if let Some((k, v)) = pair.split_once('=') {
                query.insert(k.to_string(), v.to_string());
            }
        }
        Request {
            method: "GET".into(),
            path: p.into(),
            query,
            body: String::new(),
            origin: None,
            fetch_site: None,
        }
    }

    /// Same-origin POST: no browser provenance headers, which is what the CLI and the
    /// dashboard's own fetch both look like.
    fn post(path: &str, body: &str) -> Request {
        Request {
            method: "POST".into(),
            path: path.into(),
            query: Default::default(),
            body: body.into(),
            origin: None,
            fetch_site: None,
        }
    }

    fn new_task(title: &str) -> paos_tasks::model::NewTask {
        paos_tasks::model::NewTask {
            title: title.into(),
            body: None,
            scope: "global".into(),
            org: None,
            repo: None,
            parent_id: None,
            priority: 2,
            origin: Origin::Session,
            created_by: "swift-otter".into(),
            room: None,
            start_ready: false,
        }
    }

    fn seed(w: &Web, title: &str) -> String {
        let c = w.lock();
        store::create(&c, &new_task(title), &now_iso()).unwrap()
    }

    #[test]
    fn the_board_payload_carries_the_derived_blocked_flag() {
        let w = test_web();
        let (a, b) = (seed(&w, "first"), seed(&w, "second"));
        {
            let c = w.lock();
            store::dep_add(&c, &b, &a, &now_iso()).unwrap();
        }
        let body = String::from_utf8(w.route(&get("/api/tasks")).body).unwrap();
        assert!(body.contains("\"blocked\":true"), "{body}");
        assert!(body.contains("\"blocked\":false"));
    }

    #[test]
    fn a_cross_origin_post_is_refused_like_every_other_write() {
        // The board's writes are as authoritative as /api/answer — they move work and
        // grant close-authority — so they must sit behind the same CSRF guard rather than
        // quietly being the exception.
        let w = test_web();
        let mut r = post("/api/task/state", "id=t-aaaaaa&to=done");
        r.origin = Some("https://evil.example".into());
        assert_eq!(w.route(&r).status, 403);
        let mut r = post("/api/task/grant", "id=t-aaaaaa");
        r.fetch_site = Some("cross-site".into());
        assert_eq!(w.route(&r).status, 403);
    }

    #[test]
    fn the_board_speaks_as_the_operator_so_it_can_approve_review_items() {
        // Pinning WHICH actor the dashboard is. If it spoke as a session, every operator
        // task would be unclosable from the UI; if `may_close` were skipped entirely,
        // the approval gate would exist only in the CLI.
        let w = test_web();
        let id = {
            let c = w.lock();
            let mut n = new_task("operator task");
            n.origin = Origin::Operator;
            store::create(&c, &n, &now_iso()).unwrap()
        };
        let r = w.route(&post("/api/task/state", &format!("id={id}&to=done")));
        assert_eq!(r.status, 200, "{}", String::from_utf8_lossy(&r.body));
    }

    #[test]
    fn a_refused_transition_answers_409_with_the_reason() {
        let w = test_web();
        let id = seed(&w, "x");
        {
            let c = w.lock();
            store::set_state(&c, &id, State::Dropped, &store::Actor::Operator, &now_iso())
                .unwrap();
        }
        let r = w.route(&post("/api/task/state", &format!("id={id}&to=ready")));
        assert_eq!(r.status, 409);
        assert!(String::from_utf8_lossy(&r.body).contains("dropped"));
    }

    #[test]
    fn a_comment_on_an_unowned_task_is_still_recorded_and_reports_no_delivery() {
        let w = test_web();
        let id = seed(&w, "unowned");
        let r = w.route(&post("/api/task/note", &format!("id={id}&text=hello&wake=1")));
        assert_eq!(r.status, 200);
        let body = String::from_utf8(r.body).unwrap();
        assert!(body.contains("\"delivered\":false"), "{body}");
        let c = w.lock();
        let n = query::notes(&c, &id).unwrap();
        assert!(
            n.iter().any(|x| x.text == "hello"),
            "the comment must survive even when nobody could be woken"
        );
    }

    #[test]
    fn the_badge_counts_what_needs_the_operator() {
        let w = test_web();
        let body = String::from_utf8(w.route(&get("/api/tasks")).body).unwrap();
        assert!(body.contains("\"needs_operator\":0"), "{body}");
    }

    #[test]
    fn titles_with_quotes_do_not_break_the_payload() {
        let w = test_web();
        seed(&w, "a \"quoted\" title\nwith a newline");
        let body = String::from_utf8(w.route(&get("/api/tasks")).body).unwrap();
        // The quote and the newline must arrive escaped, not raw — raw would terminate
        // the string early and the whole board would fail to parse in the browser.
        assert!(body.contains(r#"a \"quoted\" title\nwith a newline"#), "{body}");
        assert!(!body.contains("title\nwith"), "a literal newline leaked into the JSON");
    }

    #[test]
    fn the_board_assets_are_served_from_the_binary() {
        let w = test_web();
        for p in ["/tasks.css", "/tasks.js"] {
            let r = w.route(&get(p));
            assert_eq!(r.status, 200, "{p}");
            assert!(!r.body.is_empty(), "{p}");
        }
    }
}
